File: 3_Signalanalyse/MATLAB/3_Zeropadding/Zeropadding.m
% create signal (gausspulse) and its spectrum; then plot the spectrum of the zeropadded signal

clear all
close all
clc

%% PART 1
sampl_freq = 50e6;      % sampling frequency

% signal generation (gausspulse)
f_center = 1e6;         % center frequency
bandwidth = 0.6;        % relative bandwidth
tc = gauspuls('cutoff', f_center, bandwidth, [], -40);
t_center = -tc : 1/sampl_freq : tc;
y = gauspuls(t_center, f_center, bandwidth);
t = (0:length(y)-1)./sampl_freq;

% plot signal
plot(t*1e6, y)
xlabel('Time {\itt} / µs')
title(['Time domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y))])

% TODO: define frequency vector and plot absolute value of fft

% zeropadding
number_zeros = 1000;
y_padded = [y, zeros(1, number_zeros)];
t_padded = (0:length(y_padded)-1)./sampl_freq;
figure
plot(t_padded*1e6, y_padded)
xlabel('Time {\itt} / µs')
title(['Time domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y_padded))])

% TODO: define frequency vector and plot absolute value of fft

%% PART 2
% add signal with higher frequencies
c = 23;
disp(['Zusaätzliche Mittenfrequenz: ', num2str(f_center*c/1e6), ' MHz'])

y_high_freq = gauspuls(t_center, f_center*c, bandwidth/c);
y_add =y + y_high_freq;
figure
plot(t, y)
hold on
plot(t, y_high_freq, 'r')
hold on
plot(t, y_add, 'g')
xlabel('Time {\itt} / µs')
title(['Time domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y_add))])
legend('lower freq.', 'high freq.', 'sum')

% TODO: plot absolute value of fft

y_add_padded = [y_add, zeros(1, number_zeros)];

% TODO: plot absolute value of fft

File: 3_Signalanalyse/Python/3_Zeropadding/zeropadding.py
from scipy import signal
import numpy as np
import matplotlib.pyplot as plt
plt.close("all")

# PART 1
sampl_freq = 50e6       # sampling frequency

# gaussian modulated sine
f_center = 1e6          # center frequency
bandwidth = 0.6         # relative bandwidth
n = 200               # number of samples
t = np.linspace(0, n/sampl_freq, n)   # time

# without zeropadding
_, y = signal.gausspulse(t-t[-1]/2, fc=f_center, bw=bandwidth, retquad='true')

plt.figure()
plt.plot(t*1e6, y)
plt.xlabel('Time $t$ / µs')
plt.title('signal without zeropadding')

# fouriertransform
# TODO: calculate and plot absolute value of fft

# with zeropadding
number_zeros = 1000         # number of zeros padded
y_padded = np.concatenate((y, np.zeros(number_zeros)))      # padded signal
t_padded = np.linspace(0, np.size(y_padded)-1, np.size(y_padded))/sampl_freq

plt.figure()
plt.plot(t_padded*1e6, y_padded)
plt.xlabel('Time $t$ / µs')
plt.title('signal with zeropadding')

# fouriertransform
# TODO: calculate and plot absolute value of fft_padded

# PART 2
# add higher frequencies
c = 23                  # shift of center frequency
_, y_high_freq = signal.gausspulse(t-t[-1]/2, fc=f_center*c, bw=bandwidth/c, retquad='true')

# add low and high frequency part
y_add = y + y_high_freq
plt.figure()
plt.plot(t*1e6, y, 'b', label='low frequency')
plt.plot(t*1e6, y_high_freq, 'r', label='high frequency')
plt.plot(t*1e6, y_add, 'g', label='sum')
plt.xlabel('Time $t$ / µs')
plt.legend()

# TODO: calculate and plot absolute value of fft

# padded signal
y_add_padded = np.concatenate((y_add, np.zeros(number_zeros)))

# TODO: calculate and plot absolute value of fft

plt.show()
